tests/controls/gridtest.cpp

```diff
@@ -31,32 +31,165 @@
+                s += "multicell";
+            else
+                s += "inside cell";
+        }
+
+        return wxString::Format( "%s at %s", s, Coords() );
+    }
+};
+
+// Stores insertion/deletion info of rows or columns.
+struct EditInfo
+{
+    int pos, count, isRow;
+
+    EditInfo(int pos, int count, int isRow)
+        : pos(pos), count(count), isRow(isRow) { }
+};
+
+// Derive a new class inheriting from wxGrid, also to get access to its
 // protected GetCellAttr(). This is not pretty, but we don't have any other way
 // of testing this function.
 class TestableGrid : public wxGrid
 {
 public:
     explicit TestableGrid(wxWindow* parent)
-        : wxGrid(parent, wxID_ANY)
+        : wxGrid(parent, wxID_ANY), m_edit(0, 0, 0)
     {
     }
 
     wxGridCellAttr* CallGetCellAttr(int row, int col) const
     {
         return GetCellAttr(row, col);
     }
+
+    bool HasAttr(int row, int col,
+                 wxGridCellAttr::wxAttrKind kind = wxGridCellAttr::Cell) const
+    {
+        // Can't use GetCellAttr() here as it always returns an attr.
+        wxGridCellAttr* attr = GetTable()->GetAttr(row, col, kind);
+        if ( attr )
+            attr->DecRef();
+
+        return attr != NULL;
+    }
+
+    size_t GetCellAttrCount() const
+    {
+        // Note that only attributes in grid range can easily be checked
+        // and this function only counts those, not any outside of
+        // the grid (e.g. with invalid negative coords).
+        size_t count = 0;
+        for ( int row = 0; row < GetNumberRows(); ++row )
+            for ( int col = 0; col < GetNumberCols(); ++col )
+                count += HasAttr(row, col);
+
+        return count;
+    }
+
+    void SetMulticell(const Multicell& multi)
+    {
+        SetCellSize(multi.row, multi.col, multi.rows, multi.cols);
+    }
+
+    // Performs given insertions/deletions on either rows or columns.
+    void DoEdit(const EditInfo& edit);
+
+    // Returns annotated grid represented as a string.
+    wxString ToString() const;
+
+    // Used when drawing annotated grid to know what happens to it.
+    EditInfo m_edit;
+
+    // Grid as string before editing, with edit info annotated.
+    wxString m_sOriGridAnnotated;
 };
 
 } // anonymous namespace
 
+namespace Catch
+{
+
+template <> struct StringMaker<TestableGrid>
+{
+    static std::string convert(const TestableGrid& grid)
+    {
+        return ("Content before edit:\n" + grid.m_sOriGridAnnotated
+                + "\nContent after edit:\n" + grid.ToString()).ToStdString();
+    }
+};
+
+template <> struct StringMaker<Multicell>
+{
+    static std::string convert(const Multicell& multi)
+    {
+        return multi.ToString().ToStdString();
+    }
+};
+
+} // namespace Catch
+
 class GridTestCase
 {
 public:
     GridTestCase();
     ~GridTestCase();
 
 protected:
+
+    void InsertRows(int pos = 0, int count = 1)
+    {
+        m_grid->DoEdit( EditInfo(pos, count, /* isRow = */ true) );
```

Comment:
Not sure if it's worth changing things for, but we have `wxGridDirection` enum, i.e. you could use `wxGRID_ROW` here and `wxGRID_COLUMN` below instead of bool + comment.

Reply:
Nice, thanks!